dotnet/src/Azure.Iot.Operations.Connector/CloudEvents/AioCloudEventBuilder.cs

```diff
@@ -0,0 +1,253 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Azure.Iot.Operations.Connector.CloudEvents;
+
+using Services.AssetAndDeviceRegistry.Models;
+
+/// <summary>
+/// Builds AIO CloudEvents according to the message_correlation.md specification.
+/// </summary>
+public static class AioCloudEventBuilder
+{
+    /// <summary>
+    /// Builds AIO CloudEvent for a specific asset and dataset combination.
+    /// </summary>
+    /// <param name="device">ADR Device model.</param>
+    /// <param name="deviceName">Device name for fallback in source/subject generation.</param>
+    /// <param name="endpointName">Endpoint name from the asset's deviceRef.</param>
+    /// <param name="protocolSpecificIdentifier">Protocol specific device identifier.</param>
+    /// <param name="asset">ADR Asset model.</param>
+    /// <param name="dataset">ADR AssetDataset model.</param>
+    /// <param name="assetName">Asset name.</param>
+    /// <param name="messageSchemaReference">Optional message schema reference reported to ADR. Used to construct the aio-sr:// DataSchema URI.</param>
+    /// <returns>Generated AIO CloudEvent.</returns>
+    public static AioCloudEvent Build(
+        Device device,
+        string deviceName,
+        string endpointName,
+        string? protocolSpecificIdentifier,
```

Comment:
Optional params should be at the end of the param list and should default to null